File: cluster/src/main/java/com/netcompany/vertx/examples/cluster/EventBusVerticle.java
package com.netcompany.vertx.examples.cluster;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.eventbus.Message;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;

import java.net.InetAddress;
import java.net.UnknownHostException;

public class EventBusVerticle extends AbstractVerticle {
    private final Logger logger = LoggerFactory.getLogger(EventBusVerticle.class.getSimpleName());

    public final static String EXAMPLE_PUBLISH_ADDRESS = "com.netcompany.vertx.example.publish";
    public final static String EXAMPLE_SEND_ADDRESS = "com.netcompany.vertx.example.send";

    @Override
    public void start(Future<Void> startFuture) {
        JsonObject ports = config().getJsonObject("ports");
        Integer javaPort = ports != null && ports.getInteger("java") != null ? ports.getInteger("java") : 8080;

        vertx.createHttpServer().requestHandler(req -> req.response()
                .putHeader("content-type", "text/plain")
                .end("Hello from Java Vert.x!")
        ).listen(javaPort, listenRes -> {
            if (listenRes.failed()) {
                startFuture.fail(listenRes.cause());
            } else {
                setPeriodicPublisher();

                startFuture.complete();
            }
        });
    }

    private void setPeriodicPublisher() {
        try {
            JsonObject data = new JsonObject()
                    .put("testKey", "testValue")
                    .put("host", InetAddress.getLocalHost().getHostName());

            vertx.setPeriodic(10000L, aLong -> {
                logger.info("Sending publish with: " + data.encode());

                vertx.eventBus().publish(EXAMPLE_PUBLISH_ADDRESS, data);
            });

            vertx.setPeriodic(1000L, aLong -> vertx.eventBus().<JsonObject>send(EXAMPLE_SEND_ADDRESS, data, reply -> {
                if (reply.failed()) {
                    logger.error("Error in retreiving reply!", reply.cause());
                } else {
                    final Message<JsonObject> message = reply.result();

                    logger.info("Received reply from: " + message.body().getString("replier") +
                            " at " + message.address() + " with " + message.body().encode());
                }
            }));
        } catch (UnknownHostException e) {
            logger.error("Unable to build data!", e);
        }
    }
}

File: service-proxies/src/main/kotlin/com/netcompany/vertx/examples/serviceproxies/verticles/EventbusVerticle.kt
package com.netcompany.vertx.examples.serviceproxies.verticles

import com.netcompany.vertx.examples.serviceproxies.services.HeartBeatService
import com.netcompany.vertx.examples.serviceproxies.services.HeartBeatServiceImpl
import io.vertx.core.AbstractVerticle
import io.vertx.core.Future
import io.vertx.core.eventbus.MessageConsumer
import io.vertx.core.json.Json
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.Logger
import io.vertx.core.logging.LoggerFactory
import io.vertx.serviceproxy.ProxyHelper
import io.vertx.ext.web.Router
import io.vertx.ext.web.handler.CookieHandler
import io.vertx.ext.web.handler.sockjs.BridgeEventType.*
import io.vertx.ext.web.handler.sockjs.BridgeOptions
import io.vertx.ext.web.handler.sockjs.PermittedOptions
import io.vertx.ext.web.handler.sockjs.SockJSHandler
import io.vertx.kotlin.ext.web.handler.sockjs.SockJSHandlerOptions
import java.util.ArrayList




class EventbusVerticle : AbstractVerticle() {
    private val logger: Logger = LoggerFactory.getLogger(javaClass.simpleName)

    private val heartBeat = HeartBeatServiceImpl()
    private var registerService: MessageConsumer<JsonObject>? = null
    private var ebHandler: SockJSHandler? = null

    override fun start(startFuture: Future<Void>?) {
        val router: Router = Router.router(vertx)
        router.get("/").handler({ context ->
            context.response()
                    .putHeader("content-type", "text/plain")
                    .end("Hello from Java Vert.x!")
        })

        val ports = config().getJsonObject("ports")
        val javaPort = if (ports?.getInteger("java") != null) ports.getInteger("java") else 8080

        vertx.createHttpServer().requestHandler(router::accept).listen(javaPort!!) { listenRes ->
            if (listenRes.failed()) {
                startFuture?.fail(listenRes.cause())
            } else {
                deployHeartBeatService()

                deployBridge(router)

                startFuture?.complete()
            }
        }
    }

    private fun deployHeartBeatService() {
        registerService = ProxyHelper.registerService(HeartBeatService::class.java, vertx, heartBeat, "com.netcompany.heart")

        logger.info("Deployed HeartBeatService")
    }

    private fun deployBridge(router: Router) {
        val options = SockJSHandlerOptions()
                .setHeartbeatInterval(150000L)
                .setInsertJSESSIONID(false)

        ebHandler = SockJSHandler.create(vertx, options)
        ebHandler?.bridge(createBridgeOptions(), { bridgeEvent ->
            logger.debug("Event received from external client!")

            if (bridgeEvent.type() != null) {
                when (bridgeEvent.type()) {
                    SEND -> {
                        logger.debug("Send Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    PUBLISH -> {
                        logger.debug("Publish Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    RECEIVE -> {
                        logger.debug("Receive Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    REGISTER -> {
                        logger.debug("Register Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    UNREGISTER -> {
                        logger.debug("UnRegister Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    SOCKET_IDLE -> {
                        logger.debug("Socket Idle Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    SOCKET_PING -> {
                        logger.debug("Socket Ping Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    SOCKET_CLOSED -> {
                        logger.debug("Socket Closed Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    SOCKET_CREATED -> {
                        logger.debug("Socket Created Event is: " + Json.encodePrettily(bridgeEvent))

                        bridgeEvent.complete(true)
                    }

                    else -> {
                        logger.error("Unknown bridgevent!")

                        bridgeEvent.complete(true)
                    }
                }
            } else {
                logger.error("Type is null!, Message is: " + Json.encodePrettily(bridgeEvent))
            }
        })

        router.route("/eventbus/*").handler(CookieHandler.create())
        router.route("/eventbus/*").handler(ebHandler)
    }

    private fun createBridgeOptions(): BridgeOptions {
        val opts = BridgeOptions()
        opts.setInboundPermitted(createInAdresses())
        opts.setOutboundPermitted(createOutAdresses())

        return opts
    }

    private fun createInAdresses(): List<PermittedOptions> {
        val permissions = ArrayList<PermittedOptions>()
        permissions.add(PermittedOptions().setAddress("com.netcompany.heart"))

        return permissions
    }

    private fun createOutAdresses(): List<PermittedOptions> {
        val permissions = ArrayList<PermittedOptions>()
        permissions.add(PermittedOptions().setAddress("com.netcompany.heart"))

        return permissions
    }

    override fun stop(stopFuture: Future<Void>?) {
        undeployHeartBeatService()

        super.stop(stopFuture)
    }

    private fun undeployHeartBeatService() {
        ProxyHelper.unregisterService(registerService)

        logger.info("UnDeployed HeartBeatService")
    }
}